### PR TITLE
AdSizes for AdMob banner ads

### PR DESCRIPTION
**Pull request by RedBrogdon**
This PR adds a new AdSize class to the firebase_admob plugin for use in making banners (see Issue [#8098](https://github.com/flutter/flutter/issues/8098)). All the standard sizes are supported (see [this AdMob guide](https://developers.google.com/admob/android/banner#banner_sizes) for a list).

Right now, the public API and Android implementation are worked out, with iOS and finishing touches to come. I'm putting it up early so people can have a look at the API.

### CODE REVIEW

File: packages/firebase_admob/android/src/main/java/io/flutter/plugins/firebaseadmob/FirebaseAdMobPlugin.java [hunk 1]
```diff
@@ -44,7 +45,47 @@ private void callInitialize(MethodCall call, Result result) {
     result.success(Boolean.TRUE);
   }
 
-  private void callLoadAd(MobileAd ad, MethodCall call, Result result) {
+  private void callLoadBannerAd(int id, Activity activity, MethodChannel channel, MethodCall call, Result result) {
+    String adUnitId = call.argument("adUnitId");
+    if (adUnitId == null || adUnitId.isEmpty()) {
+      result.error("no_unit_id", "a non-empty adUnitId was not provided for ad id=" + id, null);
```
Comment (Phrohdoh) [inline packages/firebase_admob/android/src/main/java/io/flutter/plugins/firebaseadmob/FirebaseAdMobPlugin.java]: This wording is a bit confusing. Would you be open to changing this to something like `a null or empty adUnitId...`?

File: packages/firebase_admob/lib/firebase_admob.dart [hunk 2]
```diff
@@ -82,6 +82,65 @@ class MobileAdTargetingInfo {
   }
 }
 
+// The types of ad sizes supported for banners.
+enum AdSizeType {
+  WidthAndHeight,
+  SmartBanner,
+}
+
+/// [AdSize] represents the size of a banner ad.
```
Comment (HansMuller) [inline packages/firebase_admob/lib/firebase_admob.dart]: It would be useful to link to the relevant doc here; https://developers.google.com/admob/android/banner#banner_sizes I think

File: packages/firebase_admob/lib/firebase_admob.dart [hunk 3]
```diff
@@ -82,6 +82,65 @@ class MobileAdTargetingInfo {
...
+class AdSize {
```
Comment (HansMuller) [inline packages/firebase_admob/lib/firebase_admob.dart]: You could just expose const constructors, like `const AdSize.banner(), although it might be simpler to just expose statics, like `AdSize.banner`.

File: packages/firebase_admob/lib/firebase_admob.dart [hunk 4]
```diff
@@ -82,6 +82,65 @@ class MobileAdTargetingInfo {
...
+  final int height;
+  final int width;
+  final AdSizeType sizeType;
+
+  /// Creates a typical width-by-height banner size. In most cases, it's easier
+  /// to use one of the named constructors like [AdSize.banner].
+  AdSize({this.width, this.height})
+      : assert(width > 0),
+        assert(height > 0),
```
Comment (HansMuller) [inline packages/firebase_admob/lib/firebase_admob.dart]: I don't think you want to expose this constructor, since all of the valid ad sizes are covered by the other constructors.

It should be a (private) const constructor where all of the parameters are `@required`.

File: packages/firebase_admob/lib/firebase_admob.dart [hunk 5]
```diff
@@ -82,6 +82,65 @@ class MobileAdTargetingInfo {
...
+  AdSize.leaderboard()
+      : width = 728,
+        height = 90,
+        sizeType = AdSizeType.WidthAndHeight;
+
+  /// Creates a smart banner size. Smart banners are unique in that they will
+  /// automatically adjust their width to match that of the displaying device's
+  /// screen. Height also varies, and will be 32, 50, or 90 depending on how
+  /// tall the screen is.
+  AdSize.smartBanner()
```
Comment (HansMuller) [inline packages/firebase_admob/lib/firebase_admob.dart]: From the docs, it seems like there are only three valid height values (and you can't really specify the width?). You could either provide three constants or just provide a const constructor whose width could only be one of the 3 valid values.

Since the recommended practice is to base the banner's size on the screen size, you could also provide a `smartBannerFor(MediaQueryData)` constructor that DTRT.

File: packages/firebase_admob/lib/firebase_admob.dart [hunk 6]
```diff
@@ -161,11 +212,14 @@ class BannerAd extends MobileAd {
...
       : 'ca-app-pub-3940256099942544/2934735716';
 
+  final AdSize size;
+
   /// Create a BannerAd.
   ///
   /// A valid [adUnitId] is required.
   BannerAd({
     @required String adUnitId,
+    @required this.size,
```
Comment (HansMuller) [inline packages/firebase_admob/lib/firebase_admob.dart]: Why not give this a default value?

File: packages/firebase_admob/lib/firebase_admob.dart [hunk 7]
```diff
@@ -82,6 +82,65 @@ class MobileAdTargetingInfo {
   }
 }
 
+// The types of ad sizes supported for banners.
+enum AdSizeType {
```
Comment (HansMuller) [inline packages/firebase_admob/lib/firebase_admob.dart]: It looks like indices of these enums are significant. That's worth a '//' implementation comment.

File: packages/firebase_admob/lib/firebase_admob.dart (hunk 5 above)
Comment (RedBrogdon) [reply packages/firebase_admob/lib/firebase_admob.dart]: I think I need to revise this comment, since it clearly didn't impart the right info. :)

If you choose "smart banner" as your ad size, there are no other inputs you're allowed to give. You, as the publisher, do not determine how wide or high the ad is. The SDK will automatically set the width of the AdView to match the device width, and will set its height based on [a calculation involving device height](https://developers.google.com/admob/android/banner#smart_banners). So there's nothing more for the API to do here other than give the publisher a way to say "make me an AdSize object of the smart banner type."

File: packages/firebase_admob/lib/firebase_admob.dart (hunk 7 above)
Comment (RedBrogdon) [reply packages/firebase_admob/lib/firebase_admob.dart]: Should I just add to the comment that's there and indicate how I'm using the index values? If there's an example of a similar comment, feel free to point me to it.

File: packages/firebase_admob/lib/firebase_admob.dart (hunk 5 above)
Comment (HansMuller) [reply packages/firebase_admob/lib/firebase_admob.dart]: Sorry, I just didn't read this stuff carefully enough.

File: packages/firebase_admob/lib/firebase_admob.dart (hunk 7 above)
Comment (HansMuller) [reply packages/firebase_admob/lib/firebase_admob.dart]: https://docs.flutter.io/flutter/scheduler/SchedulerPhase-class.html uses an enum's index this way

It feels a little brittle to rely on the index because adding another id could disturb things.  On the other hand, you could easily write a regression test that checked this

File: packages/firebase_admob/android/src/main/java/io/flutter/plugins/firebaseadmob/FirebaseAdMobPlugin.java (hunk 1 above)
Comment (RedBrogdon) [reply packages/firebase_admob/android/src/main/java/io/flutter/plugins/firebaseadmob/FirebaseAdMobPlugin.java]: Done.

File: packages/firebase_admob/lib/firebase_admob.dart (hunk 2 above)
Comment (RedBrogdon) [reply packages/firebase_admob/lib/firebase_admob.dart]: Done.

File: packages/firebase_admob/lib/firebase_admob.dart (hunk 3 above)
Comment (RedBrogdon) [reply packages/firebase_admob/lib/firebase_admob.dart]: I was actually dithering about the most Dart-like way to do this, so I'm glad you brought it up. If having pre-built instances hang out as static constants is allowed, that's my preference.

File: packages/firebase_admob/lib/firebase_admob.dart (hunk 4 above)
Comment (RedBrogdon) [reply packages/firebase_admob/lib/firebase_admob.dart]: Done. If the plugin is ever expanded to handle DoubleClick for Publishers, it'll need to be made public again, but this is just fine for AdMob.

File: packages/firebase_admob/lib/firebase_admob.dart (hunk 5 above)
Comment (RedBrogdon) [reply packages/firebase_admob/lib/firebase_admob.dart]: You shouldn't have to, though, so it's good feedback.

File: packages/firebase_admob/lib/firebase_admob.dart (hunk 6 above)
Comment (RedBrogdon) [reply packages/firebase_admob/lib/firebase_admob.dart]: There are two reasons, I guess.

- Our Android and iOS APIs don't have a default, so it's what I'm used to.
- In my mind, publishers should be consciously picking a size based on what their UI looks like. Most will pick the standard 320x50 banner anyway, but I'd like for them to think at least for a moment about it.

File: packages/firebase_admob/lib/firebase_admob.dart (hunk 7 above)
Comment (RedBrogdon) [reply packages/firebase_admob/lib/firebase_admob.dart]: Makes sense. What about passing the String name of the value, then, rather than using the index?

File: packages/firebase_admob/lib/firebase_admob.dart (hunk 6 above)
Comment (HansMuller) [reply packages/firebase_admob/lib/firebase_admob.dart]: I see, sounds reasonable.

**Comment by mit-mit** [comment]
Thanks much @RedBrogdon for the contribution! Looks like there is a small merge conflict with master.  Can you take a quick look?

**Comment by Purus** [comment]
Please let me know I can be of any help on this?
Also kindly check on the issue https://github.com/flutter/flutter/issues/14222 if you can help on this.

**Comment by RedBrogdon** [comment]
The API is finished at this point, and it's implemented all the way down for iOS and Android. Tests are passing, and the code's linted.

PTAL!

**Comment by Purus** [comment]
Hope it's merged soon :)

File: packages/firebase_admob/android/src/main/java/io/flutter/plugins/firebaseadmob/FirebaseAdMobPlugin.java [hunk 8]
```diff
@@ -38,14 +39,61 @@ private FirebaseAdMobPlugin(Registrar registrar, MethodChannel channel) {
   private void callInitialize(MethodCall call, Result result) {
     String appId = call.argument("appId");
     if (appId == null || appId.isEmpty()) {
-      result.error("no_app_id", "a non-empty AdMob appId was not provided", null);
+      result.error("no_app_id", "a null or empty AdMob appId was provided", null);
```
Comment (HansMuller) [inline packages/firebase_admob/android/src/main/java/io/flutter/plugins/firebaseadmob/FirebaseAdMobPlugin.java]: NICE

**Comment by RedBrogdon** [comment]
@Purus, to answer your earlier question about what you might do to help: please give this a thorough test in whatever apps you're working on. It's playing along just fine for me, but I'm sure there are edge cases I haven't discovered.

**Comment by Purus** [comment]
Sure.. I am happy to test it in my development apps. I will test once this is merged.

**Comment by RedBrogdon** [comment]
When is this one likely to get merged?
